cassandra/cluster.py

```diff
@@ -2403,16 +2409,20 @@ def run_add_or_renew_pool():
 
             previous = self._pools.get(host)
             with self._lock:
-                while new_pool._keyspace != self.keyspace:
+                while new_pool._keyspace != self.keyspace and \
+                        not isinstance(new_pool._keyspace_exception, InvalidRequest):
                     self._lock.release()
                     set_keyspace_event = Event()
                     errors_returned = []
 
                     def callback(pool, errors):
                         errors_returned.extend(errors)
                         set_keyspace_event.set()
-
-                    new_pool._set_keyspace_for_all_conns(self.keyspace, callback)
+                    try:
```

Comment:
Issue from before, if an exception was risen here (like the connection was suddenly closed), the error risen was something like `Can't release lock if it isn't acquired`